File: Leetcode_with_Cpp/22generateParenthesis.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        dfs(res, "", 0, 0, n);
        return res;
    }
    void dfs(vector<string> &res, string str, int l, int r, int n) {
        if (l > n || r > n || r > l) return;
        if (l == n && r == n) {
            res.push_back(str);
            return;
        }
        dfs(res, str + '(', l+1, r, n);
        dfs(res, str + ')', l, r+1, n);
        return;
    }
};

File: py/22generateParenthesis.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        self.res = []
        self.func('', 0, 0, n)
        return self.res

    def func(self, str, l, r, n):
        if l == n and r == n:
            self.res.append(str)
        if l < n:
            self.func(str+'(', l+1, r, n)
        if r < l:
            self.func(str+')', l, r+1, n)